pymc_marketing/mmm/multidimensional.py

```diff
@@ -52,6 +55,92 @@
 warnings.warn(warning_msg, FutureWarning, stacklevel=1)
 
 
+class MuEffect(Protocol):
+    """Protocol for abitrary additive mu effect."""
+
+    def create_data(self, mmm: MMM) -> None:
+        """Create the required data in the model."""
+
+    def create_effect(self, mmm: MMM) -> pt.TensorVariable:
+        """Create the additive effect in the model."""
+
+
+def create_event_mu_effect(
+    df_events: pd.DataFrame,
+    prefix: str,
+    effect: EventEffect,
+) -> MuEffect:
+    """Create an event effect for the MMM.
+
+    This class has the ability to create data and mean effects for the MMM model.
+
+    """
+    if missing_columns := df_events.columns.difference(
+        ["start_date", "end_date", "name"]
+    ).tolist():
+        raise ValueError(f"Columns {missing_columns} are missing in df_events.")
+
+    effect.basis.prefix = prefix
+
+    class Effect:
+        def create_data(self, mmm: MMM) -> None:
+            model: pm.Model = mmm.model
+
+            model_dates = pd.to_datetime(model.coords["date"])
+
+            start_dates = pd.to_datetime(df_events["start_date"])
+            end_dates = pd.to_datetime(df_events["end_date"])
+
+            reference_date = model_dates.min()
+
+            model.add_coord(prefix, df_events["name"].to_numpy())
+
+            if "days" not in model:
+                pm.Data(
+                    "days",
+                    days_from_reference(model_dates, reference_date),
+                    dims="date",
+                )
+
+            pm.Data(
+                f"{prefix}_start_diff",
+                days_from_reference(start_dates, reference_date),
+                dims=prefix,
+            )
+            pm.Data(
+                f"{prefix}_end_diff",
+                days_from_reference(end_dates, reference_date),
+                dims=prefix,
+            )
+
+        def create_effect(self, mmm: MMM) -> pt.TensorVariable:
+            model: pm.Model = mmm.model
+
+            s_ref = model["days"][:, None] - model[f"{prefix}_start_diff"]
+            e_ref = model["days"][:, None] - model[f"{prefix}_end_diff"]
+
+            def create_basis_matrix(s_ref, e_ref):
+                return pt.where(
+                    (s_ref >= 0) & (e_ref <= 0),
+                    0,
+                    pt.where(pt.abs(s_ref) < pt.abs(e_ref), s_ref, e_ref),
```

Comment:
Do we want control over this? @juanitorduz

Reply:
seems reasonable

Reply:
Reasonable default or would it be passed with Basis? It causes issue if we are trying to keep track of what the user passes though for building the model

Reply:
I do not have a strong opinion so I suggest whatever you feel is easiest to maintain. We can release and collect feedback.